### DISCUSSION TITLE
Replace Color of an object in image

### DISCUSSION THREAD

**Discussion by gmek5476** [ImageMagick/ImageMagick#7024] [Help]
Hi All ,

I have an image contains a white color box.I want to color the box into Brown color leaving rest of image as it is.I tried with below command but working partially.Any help is really appreciated.
Source image 
![Source1](https://github.com/ImageMagick/ImageMagick/assets/126016862/4efacf15-5dbe-4d72-93e5-222621b284b1)

Target
![Target](https://github.com/ImageMagick/ImageMagick/assets/126016862/038e841f-e065-4821-bc88-634de1bb7b89)

Expected Output

![Expected](https://github.com/ImageMagick/ImageMagick/assets/126016862/a2986d9d-1801-4ef3-8237-470799f5d571)

The Command i tried is below
magick Source1.jpg -alpha copy +level-colors 'rgba(200,146,99,255)', Target.jpg

**Comment by snibgo** [comment]
We can get very close to the required result by matching histograms.
```
call %PICTBAT%matchHisto butcherSrc.jpg butcherReqd.jpg x.jpg
```
The result, x.jpg, is:
![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/0fa52829-2437-4f27-8e6e-67892462755f)

The Windows script matchHisto.bat is given at [Process modules](http://im.snibgo.com/customim.htm#matchHisto.bat). It needs IM to be built with my process modules, as shown on that web page.

**Comment by fmw42** [comment]
-alpha copy will not do anything.  Since the box is nearly white and the background is white, it will be hard to do that.  You will need to make a mask (threshold on white) and use it to blend the changed one with the the original using -composite.

So try this (Unix syntax)

```
magick box.jpg \
\( +clone -colorspace gray +level-colors "black,#efd8a1" \) \
\( -clone 0 -negate -threshold 0 \) \
-compose over -composite box_result.png

```

![box_result](https://github.com/ImageMagick/ImageMagick/assets/13458218/be41a0da-7662-4bc8-80fe-88024225be03)

**Reply by gmek5476** [reply]
hi Fred ,Could you please help me to correct the command.
Please find the attachment for source image ,which has soda can in a box. 
when i use the below command it applying the color for soda cans also. 
magick Source.jpg ( +clone -colorspace RGB +level-colors "black,#c99364" ) ( -clone 0 -negate -threshold 0 ) -compose over -composite Target.jpg
![Source](https://github.com/ImageMagick/ImageMagick/assets/126016862/f2a4a7a4-f8d7-481a-b740-45654c527588)
![target](https://github.com/ImageMagick/ImageMagick/assets/126016862/e4c5d820-1e39-46e5-8f96-557235e4a346)
 One more case is that when i execute the command it is default applying the brown color to background also. Please find the source and target images
![Source1](https://github.com/ImageMagick/ImageMagick/assets/126016862/b1cc420c-6e3f-46db-a307-c8b69245ba0c)
![target1](https://github.com/ImageMagick/ImageMagick/assets/126016862/e053452b-cc4b-42e1-b868-b145e62f89fe)

**Comment by fmw42** [comment]
The key is the background color and how different it is relative to your object.  The boxes are different enough in color from white.  But the soda cans have white on them and the box is white also. So it will not have a good result. You will also need to find a good fuzz value as your images are JPG and the white color may not be pure white. The fuzz value must separate the object from the background as well as account for variations in the background color.d

In the code I have replace a global threshold with a flood fill.

Box:

```
fuzzval=0
magick box.jpg \
\( -clone 0 -colorspace gray +level-colors "black,#efd8a1" \) \
\( -clone 0 -fuzz $fuzzval% -fill none -draw "alpha 0,0 floodfill" \
-background black -alpha background \
-fill white +opaque none -alpha off \) \
-compose over -composite box_result.png
```

![box_result](https://github.com/ImageMagick/ImageMagick/assets/13458218/78c919cc-1954-48af-adc6-91968f62b592)


Box2:

```
fuzzval=10
magick box2.jpg \
\( -clone 0 -colorspace gray +level-colors "black,#efd8a1" \) \
\( -clone 0 -fuzz $fuzzval% -fill none -draw "alpha 0,0 floodfill" \
-background black -alpha background \
-fill white +opaque none -alpha off \) \
-compose over -composite box2_result.png
```

![box2_result](https://github.com/ImageMagick/ImageMagick/assets/13458218/79db32ef-a14d-416f-a520-7c8db768d753)


Soda Cans:

```
fuzzval=0
magick soda.jpg \
\( -clone 0 -colorspace gray +level-colors "black,#efd8a1" \) \
\( -clone 0 -fuzz $fuzzval% -fill none -draw "alpha 0,0 floodfill" \
-background black -alpha background \
-fill white +opaque none -alpha off \) \
-compose over -composite soda_result.png
```

![soda_result](https://github.com/ImageMagick/ImageMagick/assets/13458218/9106c447-b990-4e81-af4a-277ffb11cada)
